File: sdk/typescript/aaron-client.ts
/**
 * Aaron Client — TypeScript SDK for the OPTX Network
 * ====================================================
 * Light client for interacting with the Aaron Router.
 *
 * Install:
 *   npm install @astroknots/aaron  (coming soon)
 *   // or copy this file into your project
 *
 * Usage:
 *   import { AaronClient } from './aaron-client'
 *
 *   const aaron = new AaronClient('https://astroknots.space/optx')
 *   const session = await aaron.createSession({ walletAddress: 'your-pubkey' })
 *   // Show session.qrPayload as QR code
 *   // Poll session.sessionId for "verified" status
 *
 * Docs: https://astroknots.space/docs
 */

export interface AaronSession {
  sessionId: string
  challenge: string
  expiresAt: number
  qrPayload: string
}

export interface SessionStatus {
  sessionId: string
  status: "pending" | "verified" | "expired"
  expiresAt: number
  walletAddress: string | null
  verificationId?: string
  agtWeights?: AgtWeights
}

export interface AgtWeights {
  cog: number
  emo: number
  env: number
}

export interface VerifyResult {
  status: "verified"
  verificationId: string
  walletAddress: string | null
  agtWeights: AgtWeights
  entropyScore: number
  message: string
}

export interface GazeAnalysis {
  tensor: "COG" | "EMO" | "ENV"
  confidence: number
  irisPosition: { x: number; y: number }
  timestamp: number
}

export interface HealthStatus {
  status: string
  service: string
  version: string
  timestamp: string
  active_sessions: number
}

export class AaronClient {
  private baseUrl: string
  private timeout: number

  constructor(baseUrl = "https://astroknots.space/optx", timeout = 30000) {
    this.baseUrl = baseUrl.replace(/\/$/, "")
    this.timeout = timeout
  }

  private async request<T>(
    method: "GET" | "POST",
    path: string,
    body?: Record<string, unknown>
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`
    const options: RequestInit = {
      method,
      headers: { "Content-Type": "application/json" },
      signal: AbortSignal.timeout(this.timeout),
    }
    if (body) {
      options.body = JSON.stringify(body)
    }
    const res = await fetch(url, options)
    if (!res.ok) {
      const text = await res.text()
      throw new Error(`Aaron ${method} ${path} failed (${res.status}): ${text}`)
    }
    return res.json()
  }

  /** Check Aaron Router health */
  async health(): Promise<HealthStatus> {
    return this.request("GET", "/health")
  }

  /** Create a new Jett Auth session */
  async createSession(opts?: {
    walletAddress?: string
    origin?: string
  }): Promise<AaronSession> {
    return this.request("POST", "/session", {
      wallet_address: opts?.walletAddress ?? null,
      origin: opts?.origin ?? "https://jettoptics.ai",
    })
  }

  /** Poll session status */
  async pollSession(sessionId: string): Promise<SessionStatus> {
    return this.request("GET", `/session/${sessionId}`)
  }

  /** Submit gaze proof for verification */
  async verify(proof: {
    sessionId: string
    challenge: string
    gazeSequence: string[]
    holdDurations: number[]
    polynomialEncoding: string
    verificationHash: string
    walletAddress?: string
  }): Promise<VerifyResult> {
    return this.request("POST", "/verify", {
      session_id: proof.sessionId,
      challenge: proof.challenge,
      gaze_sequence: proof.gazeSequence,
      hold_durations: proof.holdDurations,
      polynomial_encoding: proof.polynomialEncoding,
      verification_hash: proof.verificationHash,
      wallet_address: proof.walletAddress ?? null,
    })
  }

  /** Classify iris landmarks into AGT region */
  async analyzeGaze(
    irisLandmarks: Array<{ x: number; y: number; z?: number }>,
    timestamp: number
  ): Promise<GazeAnalysis> {
    return this.request("POST", "/gaze/analyze", {
      iris_landmarks: irisLandmarks,
      timestamp,
    })
  }

  /**
   * Wait for session to be verified (polling helper).
   * Resolves when status is "verified", rejects on "expired" or timeout.
   */
  async waitForVerification(
    sessionId: string,
    pollIntervalMs = 2000,
    timeoutMs = 120000
  ): Promise<SessionStatus> {
    const start = Date.now()
    while (Date.now() - start < timeoutMs) {
      const status = await this.pollSession(sessionId)
      if (status.status === "verified") return status
      if (status.status === "expired") throw new Error("Session expired")
      await new Promise((r) => setTimeout(r, pollIntervalMs))
    }
    throw new Error("Verification timed out")
  }
}

File: sdk/python/aaron_client.py
"""
Aaron Client — Python SDK for the OPTX Network
================================================
Light client for interacting with the Aaron Router.
No internal AGT logic — just HTTP calls to the router.

Install:
    pip install requests

Usage:
    from aaron_client import AaronClient

    client = AaronClient("https://astroknots.space/optx")
    session = client.create_session(wallet_address="your-solana-pubkey")
    print(session["qrPayload"])  # Show as QR code

    # After MOJO app submits gaze proof:
    status = client.poll_session(session["sessionId"])
    print(status["status"])  # "verified"

Docs: https://astroknots.space/docs
"""

import requests
from typing import Dict, Any, Optional, List


class AaronClient:
    """Client for the Aaron Router — OPTX network authentication."""

    def __init__(self, base_url: str = "https://astroknots.space/optx"):
        self.base_url = base_url.rstrip("/")
        self.timeout = 30

    def health(self) -> Dict[str, Any]:
        """Check Aaron Router health."""
        resp = requests.get(f"{self.base_url}/health", timeout=self.timeout)
        resp.raise_for_status()
        return resp.json()

    def create_session(
        self,
        wallet_address: Optional[str] = None,
        origin: str = "https://jettoptics.ai",
    ) -> Dict[str, Any]:
        """
        Create a new Jett Auth session.

        Returns:
            {
                "sessionId": "abc123...",
                "challenge": "hex-challenge...",
                "expiresAt": 1234567890000,
                "qrPayload": "{...}"  # JSON string for QR code
            }
        """
        payload = {"wallet_address": wallet_address, "origin": origin}
        resp = requests.post(
            f"{self.base_url}/session", json=payload, timeout=self.timeout
        )
        resp.raise_for_status()
        return resp.json()

    def poll_session(self, session_id: str) -> Dict[str, Any]:
        """
        Poll session status.

        Returns:
            {
                "sessionId": "...",
                "status": "pending" | "verified" | "expired",
                "agtWeights": {"cog": 0.33, "emo": 0.33, "env": 0.33},  # if verified
                "verificationId": "..."  # if verified
            }
        """
        resp = requests.get(
            f"{self.base_url}/session/{session_id}", timeout=self.timeout
        )
        resp.raise_for_status()
        return resp.json()

    def verify(
        self,
        session_id: str,
        challenge: str,
        gaze_sequence: List[str],
        hold_durations: List[int],
        polynomial_encoding: str,
        verification_hash: str,
        wallet_address: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Submit gaze proof for verification.

        Args:
            session_id: From create_session response
            challenge: From create_session response
            gaze_sequence: ["COG", "EMO", "ENV", "COG", "EMO", "ENV"]
            hold_durations: [650, 700, 550, 600, 680, 720] (ms per position)
            polynomial_encoding: "132123" (1=COG, 2=EMO, 3=ENV)
            verification_hash: SHA-256 of gaze data
            wallet_address: Solana wallet pubkey

        Returns:
            {
                "status": "verified",
                "verificationId": "...",
                "agtWeights": {"cog": 0.33, "emo": 0.33, "env": 0.33},
                "entropyScore": 1584,
                "message": "Gaze proof accepted."
            }
        """
        payload = {
            "session_id": session_id,
            "challenge": challenge,
            "gaze_sequence": gaze_sequence,
            "hold_durations": hold_durations,
            "polynomial_encoding": polynomial_encoding,
            "verification_hash": verification_hash,
            "wallet_address": wallet_address,
        }
        resp = requests.post(
            f"{self.base_url}/verify", json=payload, timeout=self.timeout
        )
        resp.raise_for_status()
        return resp.json()

    def analyze_gaze(
        self,
        iris_landmarks: List[Dict[str, float]],
        timestamp: float,
    ) -> Dict[str, Any]:
        """
        Classify iris landmarks into AGT region (COG/EMO/ENV).

        Args:
            iris_landmarks: [{x, y, z}, ...] from MediaPipe FaceLandmarker
            timestamp: Unix timestamp

        Returns:
            {
                "tensor": "COG" | "EMO" | "ENV",
                "confidence": 0.85,
                "irisPosition": {"x": 0.5, "y": 0.3}
            }
        """
        payload = {"iris_landmarks": iris_landmarks, "timestamp": timestamp}
        resp = requests.post(
            f"{self.base_url}/gaze/analyze", json=payload, timeout=self.timeout
        )
        resp.raise_for_status()
        return resp.json()


# ─── Convenience usage ────────────────────────────────────────────────────────
if __name__ == "__main__":
    client = AaronClient()
    print("Health:", client.health())
